src/util-ja3.c

```diff
@@ -0,0 +1,289 @@
+/* Copyright (C) 2007-2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Mats Klepsland <mats.klepsland@gmail.com>
+ *
+ * Functions used to generate JA3 fingerprint.
+ */
+
+#include "suricata-common.h"
+
+#include "util-ja3.h"
+
+#define MD5_STRING_LENGTH 33
+
+/**
+ * \brief Allocate new buffer.
+ *
+ * \return pointer to buffer on success.
+ * \return NULL on failure.
+ */
+JA3Buffer *Ja3BufferInit(void)
+{
+    JA3Buffer *buffer = SCCalloc(1, sizeof(JA3Buffer));
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error allocating memory for JA3 buffer");
+        return NULL;
+    }
+
+    return buffer;
+}
+
+/**
+ * \brief Free allocated buffer.
+ *
+ * \param buffer The buffer to free.
+ */
+void Ja3BufferFree(JA3Buffer *buffer)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be NULL");
+        return;
+    }
+
+    if (buffer->data != NULL) {
+        SCFree(buffer->data);
+    }
+
+    SCFree(buffer);
+}
+
+/**
+ * \internal
+ * \brief Resize buffer if it is full.
+ *
+ * \param buffer The buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+static int Ja3BufferResizeIfFull(JA3Buffer *buffer, unsigned len)
```

Comment:
please use explicit type instead of just 'unsigned', so something like uint32_t

Reply:
Will do.